### ISSUE TITLE
Traefik SPI Provider

### ISSUE THREAD

**Issue by rtribotte** [keycloak/keycloak#17171]
### Description

This feature proposal is to add an SPI provider specific to Traefik, just like there are specific ones for Apache, HAPROXY, and NGINX.

It would support specificities of the header value produced by the [PassTLSClientCert middleware](https://doc.traefik.io/traefik/middlewares/http/passtlsclientcert/) containing the client certificate.

### Discussion

A related issue on Traefik repository: https://github.com/traefik/traefik/issues/9669

### Motivation

On the Traefik project, we have seen quite a few community use cases involving Keycloak as an Authorization/Authentication server.

And yet, today, there is no SPI provider implementation specific to Traefik available in Keycloak.
For this reason, some of our users were relying on the NGINX SPI provider, which was unintentionally compatible with the header produced by the PassTLSClientCert middleware in Traefik, which contains the client certificate.

Since v2.9.2, the PassTLSClientCert middleware behavior changed and the client certificate PEM bytes are not URL encoded anymore.

Unfortunately, this new behavior makes the NGINX SPI provider unusable with Traefik v2.9.2+, and constitutes a blocker to upgrade for some users of the Traefik community.

After some investigation, it seems that no available SPI provider that can fit with Traefik v2.9.2+ and that can be an alternative to the NGINX one, so a Traefik dedicated SPI provider is very much needed.

### Details

I'm opening this issue to get approval from the Keycloak maintainers to move forward with opening a PR (which is already ready) with the suggested changes.

**Comment by ahus1** [comment]
Thank you for investigating this issue. While I'm not a maintainer, but only a Keycloak member, I have the following comment: 

Keycloak has the feature of providers. By looking at the APIs involved here, it has been quite stable over the years. So you would be possible to have this as a separate GitHub project that provides it own tests and documentation on how to use it and would be community maintained. Users would download the JAR (maybe by a combination of Keycloak/Traefik server version), place it in their provider folder, and provide command line options as needed. Such a project bring together the people with the knowledge of both Traefik and Keycloak.

Would that work for you? I'm not sure if you want to host this just by yourself, or as a contrib repository in an organization of GitHub.

**Comment by rtribotte** [comment]
Hello @ahus1,

Thank you for your feedback, and your suggestion (which indeed seems doable, but would complicate things a little bit for Traefik users).

However, we are a bit confused about the reason for your suggestion. We do not understand why you're talking about the stability of the APIs. How would including the Traefik provider affect in any way the APIs of Keycloak itself?
As an open-source project ourselves, we perfectly understand the concern about maintenance costs, but it does not seem like including the Traefik provider would add a big burden on the Keycloak project.

**Comment by ahus1** [comment]
> However, we are a bit confused about the reason for your suggestion. We do not understand why you're talking about the stability of the APIs. How would including the Traefik provider affect in any way the APIs of Keycloak itself?

@rtribotte - please let me rephrase: The API for extension in Keycloak used in this case has been quite stable over the years, meaning there haven't been changes that would make an extension incompatible with a new version of Keycloak. 

This simplifies things for people who want to provide an extension that builds on top of this extension, as a single version of the extension can cover multiple Keycloak versions. 

Adding the code for provider wouldn't affect the stability of Keycloak's core functionality. Looking at the existing providers, they [have only minimal docs](https://www.keycloak.org/server/reverseproxy), and AFAIK no integration tests. 

If you want to pull in a maintainer to this discussion, please send a short note to the developer mailing list: https://www.keycloak.org/community

**Comment by rtribotte** [comment]
@ahus1 Thanks for the clarification.

As you suggested, I left a message on the developer mailing list.

**Comment by schnerring** [comment]
Is there a workaround to make Traefik v2.9.4+ work? Are all existing providers incompatible?

**Comment by rtribotte** [comment]
Hello @schnerring,

> Is there a workaround to make Traefik v2.9.4+ work? Are all existing providers incompatible?
Yes, that was my conclusion at the time of opening this issue.

But, without being able to confirm this, the comment https://github.com/traefik/traefik/issues/9669#issuecomment-1404820473 suggests that the Apache SPI provider might work with Traefik v2.9.4+.
Did you have a chance to test it?

**Comment by XeroByt3** [comment]
Just gonna add over here that as far as I can tell, none of the providers are compatible with traefik, so it would be awesome to get a dedicated SPI provider specific to Traefik 😀

- NGINX doesn't work due to the encoding difference
- Apache doesn't work due to the way the client chain is sent [mentioned here](https://github.com/traefik/traefik/issues/9669#issuecomment-1430066825).
- HAProxy doesn't work - I don't know why.

**Comment by ff05** [comment]
Hi,
what is the status here?
Will the Traefik Provider be implemented?

**Comment by teamosceola** [comment]
Are there any new updates on this?
